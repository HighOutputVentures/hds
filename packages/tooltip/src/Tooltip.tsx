import {
  Box,
  PlacementWithLogical,
  Text,
  Tooltip as ChakraTooltip,
} from "@chakra-ui/react";
import * as React from "react";
import invariant from "tiny-invariant";

type ColorScheme = "dark" | "light";

export interface TooltipProps {
  label: string;
  children: React.ReactNode;
  hasArrow?: boolean;
  placement?: PlacementWithLogical;
  colorScheme?: ColorScheme;
  supportingText?: string;
}

export default function Tooltip(props: TooltipProps) {
  const {
    label,
    hasArrow,
    placement,
    colorScheme = "dark",
    supportingText,
    children,
  } = props;

  const LabelWithSupportingText = () => (
    <Box
      maxWidth={"320px"}
      minHeight="34px"
      padding={"12px"}
      rounded={"8px"}
      width={"full"}
    >
      <Text
        fontWeight={"500"}
        fontSize="12px"
        lineHeight={"18px"}
        marginBottom="4px"
        color={"neutrals.100"}
        {...(colorScheme === "light" && {
          color: "neutrals.700",
        })}
        data-testid="hds.tooltip.label"
      >
        {label}
      </Text>

      <Text
        color={"neutrals.300"}
        fontSize="12px"
        fontWeight={"400"}
        lineHeight={"18px"}
        {...(colorScheme === "light" && {
          color: "neutrals.600",
        })}
        data-testid="hds.tooltip.supporting-text"
      >
        {supportingText}
      </Text>
    </Box>
  );

  invariant(
    React.isValidElement(children),
    "'Children' must be a valid React element",
  );

  return (
    <ChakraTooltip
      label={!supportingText ? label : <LabelWithSupportingText />}
      hasArrow={hasArrow}
      placement={placement}
      color={"neutrals.100"}
      bgColor={"neutrals.900"}
      rounded={"8px"}
      {...(colorScheme === "light" && {
        color: "neutrals.700",
        bgColor: "white",
        boxShadow: lightThemeBoxShadow,
      })}
      {...(!supportingText && {
        paddingY: "8px",
        paddingX: "12px",
      })}
      data-testid="hds.tooltip"
    >
      {React.cloneElement<any>(children, {
        "data-testid": "hds.tooltip.trigger",
      })}
    </ChakraTooltip>
  );
}

const lightThemeBoxShadow = [
  "0px 12px 16px -4px rgba(16, 24, 40, 0.08)",
  "0px 4px 6px -2px rgba(16, 24, 40, 0.03)",
].join();
