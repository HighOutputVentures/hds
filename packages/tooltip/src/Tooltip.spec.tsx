import { ChakraProvider, Icon } from '@chakra-ui/react';
import '@testing-library/jest-dom';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import { Tooltip } from './Tooltip';

describe('Checkbox Component', () => {
  beforeEach(() => {
    render(
      <ChakraProvider>
        <Tooltip label="This is a tooltip">
          <Icon />
        </Tooltip>
      </ChakraProvider>,
    );
  });

  it('should toggle tooltip when hover and unhover tooltip trigger', async () => {
    const tooltip_child = screen.getByTestId('hds.tooltip.trigger');

    await act(async () => {
      fireEvent.mouseOver(tooltip_child);
    });

    waitFor(() => {
      const tooltip = screen.getByTestId('hds.tooltip');
      expect(tooltip).toBeInTheDocument();
    });

    await act(async () => {
      fireEvent.mouseOver(tooltip_child);
    });

    await waitFor(() => {
      const tooltip = screen.queryByText(/This is a tooltip/);

      expect(tooltip).not.toBeInTheDocument();
    });
  });

  it('Should match snapshot', () => {
    const component = renderer.create(
      <ChakraProvider>
        <Tooltip label="This is a tooltip">
          <Icon />
        </Tooltip>
      </ChakraProvider>,
    );

    const snapshot = component.toJSON();
    expect(snapshot).toMatchSnapshot();
    component.unmount();
  });
});
