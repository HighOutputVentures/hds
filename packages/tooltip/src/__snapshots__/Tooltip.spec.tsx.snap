// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Component Should match snapshot 1`] = `
[
  <svg
    className="chakra-icon css-onkibi"
    data-testid="hds.tooltip.trigger"
    focusable={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onPointerDown={[Function]}
    onPointerEnter={[Function]}
    viewBox="0 0 24 24"
  >
    <g
      stroke="currentColor"
      strokeWidth="1.5"
    >
      <path
        d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
        fill="none"
        strokeLinecap="round"
      />
      <path
        d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
        fill="currentColor"
        strokeLinecap="round"
      />
      <circle
        cx="12"
        cy="12"
        fill="none"
        r="11.25"
        strokeMiterlimit="10"
      />
    </g>
  </svg>,
  <span />,
  <span
    hidden={true}
    id="__chakra_env"
  />,
]
`;
